oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -1,282 +1,164 @@
     override fun offerSimulationResult(request: SimulationResponseDTO, responseObserver: StreamObserver<SimulationResultConfirmDTO>) = responseObserver.consume {
-        val output = modelService.simulations.getNamed(request.name)?.output
-        if (output != null) {
-            output.send(request)
-        } else {
-            throw IllegalStateException("no simulation '${request.name}' or buffer full")
-        }
-        SimulationResultConfirmDTO.newBuilder().build()
+        modelService.simulations.getNamed(request.name)?.output.sendAndRespond(request) { SimulationResultConfirmDTO.newBuilder().build() }
     }
 
     override fun offerErrorResult(request: ErrorResponseDTO, responseObserver: StreamObserver<ErrorConfirmDTO>) = responseObserver.consume {
-        val error = modelService.simulations.getNamed(request.name)?.error
-        if (error != null) {
-            error.send(request)
-        } else {
-            throw IllegalStateException("no simulation '${request.name}' or buffer full")
-        }
-        ErrorConfirmDTO.newBuilder().build()
+        modelService.simulations.getNamed(request.name)?.error.sendAndRespond(request) { ErrorConfirmDTO.newBuilder().build() }
     }
 
     override fun offerSimulationConfig(request: NodeStatusCommandOrResponseDTO, responseObserver: StreamObserver<NodeChangeConfirmDTO>) = responseObserver.consume {
-        val update = modelService.simulations.getNamed(request.name)?.update
-
-        if (update != null) {
-            update.send(request)
-        } else {
-            throw IllegalStateException("no simulation '${request.name}' or buffer full")
-        }
-        NodeChangeConfirmDTO.newBuilder().build()
-    }
-
-    override fun updateNode(request: NodeStatusCommandOrResponseDTO, responseObserver: StreamObserver<NodeChangeConfirmDTO>) = responseObserver.consume {
-        val newNode = updateFromResponse(request)
-        modelService.updateSim(newNode)
-        eventBus.post(StatusUpdateEvent("${request.name} updated"))
-        NodeChangeConfirmDTO.newBuilder().setMessage("Simulation updated with inputs: ${newNode.inputs} outputs: ${newNode.outputs}").build()
+        modelService.simulations.getNamed(request.name)?.update.sendAndRespond(request) { NodeChangeConfirmDTO.newBuilder().build() }
     }
 
-
-    override fun sendMessage(request: MessageCommandDTO, responseObserver: StreamObserver<MessageReponseDTO>) = responseObserver.consume {
-        eventBus.post(NewMessageEvent(Message(request.name, request.message)))
-        MessageReponseDTO.newBuilder().build()
-    }
-
-    override fun unregister(request: UnRegistrationRequestDTO, responseObserver: StreamObserver<UnRegistrationResponseDTO>) = responseObserver.consume {
-        val unregistered = modelService.closeSim(request.name)
-        UnRegistrationResponseDTO.newBuilder().setUnregistered(unregistered).build()
+    private suspend fun <T, U> Channel<T>?.sendAndRespond(request : T, responseBuilder : () -> U) : U {
```

Comment:
This doesn't add too much, I was just trying to remove these duplicated code but I wasn't good enough to write a generic function to refactor this. However, in 12_b when I was doing testing, the helper extension with GlobalScope.launch give me harder problem with unit test.

Reply:
yeah so as per our chat, the problem is GRPC. You can write tests like this:

```kt
//setup
val endpointUnderTest = OptimizerEndpoint(dep1, dep2)
val startMessage = StartOptimizationDTOBuilder....build()
val responseObserved = CompletableDeferred<OptimizationResponseDTO>()
val responseObserver = object: StreamObserver<OptimizationResponseDTO>{
  private var value: OptimizationResponseDTO? = null
  override fun onNext(message: OptimizationResponseDTO) { value = message }
  override fun onError(ex: Throwable) { responseObserved.completeExceptionally(ex) }
  override fun onComplete() { responseObserved.complete(value ?: throw NoSuchElementException()) }
}

//act
endpointUnderTest.startOptimization(startMessage, responseObserver)
val outputMessage = responseObserved.await()

//assert
assertThat(outputMessage).isEqualTo(OptimizationResponseDTOBuilder...build())
```

you can dress that up with a few helper functions too if you want.